### PR TITLE
Final vars

### PR DESCRIPTION
**Pull request by spookydonut**
adds final vars, `var/final/foo` similar to how `var/static/foo` etc work.

This also checks for duplicate definitions of the same var eg;
```
/a
  var/foo
/a/b
  var/foo
```
will now raise an error in dreamchecker.

### CODE REVIEW

File: src/dreamchecker/README.md [hunk 1]
```diff
@@ -52,9 +52,11 @@ be enabled:
 #ifdef SPACEMAN_DMM
 	#define RETURN_TYPE(X) set SpacemanDMM_return_type = X
 	#define SHOULD_CALL_PARENT(X) set SpacemanDMM_should_call_parent = X
+  #define FINAL_VAR(X) var/final/X
 #else
 	#define RETURN_TYPE(X)
 	#define SHOULD_CALL_PARENT(X)
+  #define FINAL_VAR(X) var/X
```
Comment (SpaceManiac) [inline src/dreamchecker/README.md]: Do these macros work as written with types, e.g. `FINAL_VAR(datum/reagents/reagents)`?

Maybe something like `#define FINAL_VAR var/final` or `var` then you can write `FINAL_VAR/whatever` would work better. I want to get the ergonomics on this right as far as both typed and untyped vars go.

Comment (spookydonut) [reply src/dreamchecker/README.md]: They get detected the same way global/static/tmp etc get detected by the parser, so yes

Comment (spookydonut) [reply src/dreamchecker/README.md]: `FINAL_VAR/whatever` might work better yeah

File: src/dreammaker/objtree.rs [hunk 2]
```diff
@@ -890,12 +909,14 @@ impl ObjectTree {
             is_static,
             is_const,
             is_tmp,
+            is_final,
             type_path,
         };
         var_type.suffix(&suffix);
 
-        let symbols = &mut self.symbols;
+        let symbolid = self.symbols.allocate();
```
Comment (SpaceManiac) [inline src/dreammaker/objtree.rs]: Old code allocates a symbol ID only if `is_declaration` is true and there wasn't already an entry, now it's unconditional. Seems like an undesirable change to me.

File: src/dreammaker/objtree.rs [hunk 3]
```diff
@@ -1033,6 +1055,55 @@ impl ObjectTree {
         self.register_proc(context, location, parent, proc_name, declaration, parameters, code)
     }
 
+    pub fn check_var_defs(&self, context: &Context) {
```
Comment (SpaceManiac) [inline src/dreammaker/objtree.rs]: Is this using any private info or can it go in `dreamchecker/lib.rs` easily?

File: src/dreammaker/objtree.rs [hunk 4]
```diff
@@ -71,6 +71,7 @@ pub struct VarValue {
 #[derive(Debug, Clone)]
 pub struct TypeVar {
     pub value: VarValue,
+    pub location: Location,
```
Comment (SpaceManiac) [inline src/dreammaker/objtree.rs]: Is this necessary? `.value.location` should be the same.

File: src/dreamchecker/lib.rs [hunk 5]
```diff
@@ -492,6 +493,58 @@ fn error<S: Into<String>>(location: Location, desc: S) -> DMError {
...
+            else { continue });
+
+        guard!(let Some(typeref) = objtree.find(path)
+            else { continue });
+
+        for parent_typeref in typeref.iter_parent_types() {
+            if parent_typeref.is_root() {
+                break;
+            }
+            guard!(let parent = parent_typeref.get()
```
Comment (SpaceManiac) [inline src/dreamchecker/lib.rs]: this guard isn't actually protecting anything

**Comment by SpaceManiac** [comment]
Added some touchups on top of this
https://github.com/SpaceManiac/SpacemanDMM/compare/2b8598b...d490e3a

**Pull request merged**